### ISSUE TITLE
Reverb Effect performance is very bad on low-spec machines

### ISSUE THREAD

**Issue by gl326** [YoYoGames/GameMaker-Bugs#6674]
### Description

We recently put out a public demo of our game with reverb effects added to character foley while they're indoors. Unfortunately, we received an overwhelming number of complaints that while reverb effects were active, many users saw a massive lag / drop in performance. Although we don't have comprehensive hardware diagnostics for users who reported this issue, our best guess at a common element was that users were on relatively low-spec devices, and we were later able to replicate performance issues by testing the game on an old laptop.

I have sent source code for the project privately for testing. 

[reverb-effect-performance-package.zip](https://github.com/user-attachments/files/16228959/reverb-effect-performance-package.zip)

### Expected Change

We hope reverb audio effects can have their performance improved to the point that we are confident enabling them for users by default. We would accept having an alterative lower-intensity effect option which offers better performance for users on low-spec devices.

### Steps To Reproduce

[file_John Boyd_LaggyScene.txt](https://github.com/user-attachments/files/16228918/file_John.Boyd_LaggyScene.txt)

1. Download the included save file and delete its extension (it should be extensionless, but was edited to be a .txt file for the sake of upload).
2. Place the included save file in AppData/Local/beastieball/save/
3. Run our project in debug mode
4. You can press ~ to access a debug menu. On the final tab called "Load," if the save file is in the correct location you should see it available to load. Load the save file and you will warp to a location where users have reported seeing lag.
5. Observe performance in this area while walking around or interacting with characters. On our dev machines we see no issue, but low-spec devices will see a huge performance loss in this scene due to reverb being applied to one of the busses. 
6. Press P to open the in-game pause menu, then navigate to Settings > Performance > Audio Effects. You should observe the difference in performance when enabling/disabling audio effects.

### How reliably can you recreate this issue using your steps above?

Always

### Which version of GameMaker are you reporting this issue for?

2024.6.1 (Monthly)

### Which platform(s) are you seeing the problem on?

Windows

### Contact Us Package Attached?

- [X] I have attached my Contact Us Package

### Sample Project Added?

- [ ] I have included a small sample project

**Comment by toby-yoyo** [comment]
Although we weren't able to reproduce any visual signs of performance impact here, we did do an analysis of the effect and made changes based on the findings from that:
- Fixed an issue where instances of the reverb effect could suddenly experience a significant deterioration in performance due to their internal filters generating subnormal numbers. I believe that this was likely the main cause of this issue and we were able to reproduce it in the problematic area in-game. Affected instances were taking up to 10x longer to compute than normal.
- Reduced the time taken to compute the effect more generally. In a test environment, this went down from an average of ~128ns/frame (stereo) to ~77ns/frame, which is a saving of ~40%.
- Reduced the memory usage of the effect. Most of the reduction will come from deallocating buffers of internal filters that are on unused channels, so this should be a significant saving for those using stereo devices.
- Slightly reduced the amount of work done to instantiate the effect and to update its parameters.
